core/reorder/nested_dissection.cpp

```diff
@@ -0,0 +1,200 @@
+SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT
+LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES; LOSS OF USE,
+DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON ANY
+THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT
+(INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE
+OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+******************************<GINKGO LICENSE>*******************************/
+
+#include <ginkgo/core/reorder/nested_dissection.hpp>
+
+
+#include <memory>
+
+
+#include <ginkgo/core/base/executor.hpp>
+#include <ginkgo/core/matrix/sparsity_csr.hpp>
+
+
+#if GKO_HAVE_METIS
+#include GKO_METIS_HEADER
+#endif
+
+
+#include <ginkgo/core/base/temporary_clone.hpp>
+
+
+#include "core/base/allocator.hpp"
+
+
+namespace gko {
+namespace experimental {
+namespace reorder {
+namespace {
+
+
+std::string metis_error_message(idx_t metis_error)
+{
+    switch (metis_error) {
+    case METIS_ERROR_INPUT:
+        return "METIS_ERROR_INPUT";
+    case METIS_ERROR_MEMORY:
+        return "METIS_ERROR_MEMORY";
+    case METIS_ERROR:
+        return "METIS_ERROR";
+    case METIS_OK:
+        return "METIS_OK";
+    default:
+        return "<unknown>";
+    }
+}
+
+
+std::array<idx_t, METIS_NOPTIONS> build_metis_options(
+    const std::unordered_map<int, int>& options)
+{
+    std::array<idx_t, METIS_NOPTIONS> result{};
+    METIS_SetDefaultOptions(result.data());
+    for (auto pair : options) {
+        if (pair.first < 0 || pair.first >= METIS_NOPTIONS) {
+            throw MetisError(__FILE__, __LINE__, "build_metis_options",
+                             "Invalid option ID " + std::to_string(pair.first));
+        }
+        result[pair.first] = pair.second;
+    }
+    // make sure users don't accidentally switch on 1-based indexing
+    result[METIS_OPTION_NUMBERING] = 0;
+    return result;
+}
+
+
+template <typename IndexType>
+void metis_nd(std::shared_ptr<const Executor> host_exec, size_type num_rows,
+              const IndexType* row_ptrs, const IndexType* col_idxs,
+              const std::array<idx_t, METIS_NOPTIONS>& options, IndexType* perm,
+              IndexType* iperm)
+{
+    auto nvtxs = static_cast<idx_t>(num_rows);
+    vector<idx_t> tmp_row_ptrs(row_ptrs, row_ptrs + num_rows + 1, {host_exec});
+    vector<idx_t> tmp_col_idxs(col_idxs, col_idxs + row_ptrs[num_rows],
+                               {host_exec});
+    vector<idx_t> tmp_perm(num_rows, {host_exec});
+    vector<idx_t> tmp_iperm(num_rows, {host_exec});
+    auto result = METIS_NodeND(&nvtxs, tmp_row_ptrs.data(), tmp_col_idxs.data(),
+                               nullptr, const_cast<idx_t*>(options.data()),
+                               tmp_perm.data(), tmp_iperm.data());
+    if (result != METIS_OK) {
+        throw MetisError(__FILE__, __LINE__, "METIS_NodeND",
+                         metis_error_message(result));
+    }
```

Comment:
replaced by metis_nd<idx_t>?

Reply:
The __LINE__ info should be sufficient to differentiate the two cases, and METIS_NodeND is likely more familiar to users. What do you think?

Reply:
ah, my statement leads confusion,
I mean line 112-118 can call the following `metis_nd<idx_t>`.  but it is just a nit

Reply:
ah right thanks, I'll see if this works easily.